orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -691,33 +722,362 @@ Status TransformGraphForPipeline(
+    std::vector<NodeArg*> send_input_args;
+    std::vector<NodeArg*> send_output_args;
+    std::vector<NodeArg*> recv_input_args;
+    std::vector<NodeArg*> recv_output_args;
+
+    auto cut_index_str = std::to_string(index);
+    // add input node_arg and initializer for send/recv
+    AddNewNodeArgAndInitializer<bool>(graph,
+                                      "send_input_signal" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                      true, /* initializer data */
+                                      send_input_args,
+                                      new_input_names);
+    AddNewNodeArgAndInitializer<bool>(graph,
+                                      "recv_input_signal" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                      true, /* initializer data */
+                                      recv_input_args,
+                                      new_input_names);
+
+    AddNewNodeArgAndInitializer<size_t>(graph,
+                                      "send_dst_rank" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                      index + 1, /* initializer data */
+                                      send_input_args,
+                                      new_input_names);
+    AddNewNodeArgAndInitializer<size_t>(graph,
+                                      "recv_src_rank" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                      index, /* initializer data */
+                                      recv_input_args,
+                                      new_input_names);
+    // add output node_arg for send/recv
+    AddNewNodeArg(graph, "send_output_signal" + cut_index_str, ONNX_NAMESPACE::TensorProto_DataType_BOOL, send_output_args, new_output_names);
+    AddNewNodeArg(graph, "receive_output_signal" + cut_index_str, ONNX_NAMESPACE::TensorProto_DataType_BOOL, recv_output_args, new_output_names);
+
+    // add attribute data for send/recv
+    ONNX_NAMESPACE::AttributeProto tag;
+    tag.set_name("tag");
+    tag.set_type(ONNX_NAMESPACE::AttributeProto_AttributeType::AttributeProto_AttributeType_INT);
+    // currently hard-coded all tag to be 0. May need to change when multiple GPU stream is used.
+    tag.set_i(static_cast<int64_t>(0));
+
+    ONNX_NAMESPACE::AttributeProto element_types;
+    element_types.set_name("element_types");
+    element_types.set_type(ONNX_NAMESPACE::AttributeProto_AttributeType::AttributeProto_AttributeType_INTS);
+
+    // for each edge in this group, perform edge cut
+    for (auto& id : edgeIds) {
+      // find node whose output contains id.node_arg_name
+      auto producer_node = graph.GetMutableProducerNode(id.node_arg_name);
+      if (producer_node == nullptr) {
```

Comment:
>producer_node == nullptr [](start = 10, length = 24)

if (!producer_node)

Reply:
I think basiclly they are the same. I have no problem changing it. Just curious if there is any guidance on prefer one over the other.

---
In reply to: [429526366](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r429526366) [](ancestors = 429526366)

Reply:
It's shorter and more readable if people has smaller screen. :)

---
In reply to: [430585524](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r430585524) [](ancestors = 430585524,429526366)